src/hotspot/share/prims/jvmtiThreadState.inline.hpp

```diff
@@ -78,18 +78,22 @@ inline JvmtiThreadState* JvmtiThreadState::state_for_while_locked(JavaThread *th
 
   NoSafepointVerifier nsv;  // oop is safe to use.
 
-  if (thread_oop == nullptr) {  // Then thread should not be null (see assert above).
-    thread_oop = thread->jvmti_vthread() != nullptr ? thread->jvmti_vthread() : thread->threadObj();
-  }
-
   // In a case of unmounted virtual thread the thread can be null.
   JvmtiThreadState *state = thread == nullptr ? nullptr : thread->jvmti_thread_state();
 
   if (state == nullptr && thread != nullptr && thread->is_exiting()) {
     // Don't add a JvmtiThreadState to a thread that is exiting.
     return nullptr;
   }
-  if (state == nullptr || state->get_thread_oop() != thread_oop) {
+
+  if (thread_oop == nullptr) {  // Then thread should not be null (see assert above).
+    thread_oop = thread->jvmti_vthread() != nullptr ? thread->jvmti_vthread() : thread->threadObj();
+  }
+  // The state->get_thread_oop() may be null if the state is created during
+  // the allocation of the thread oop when a native thread is attaching. Make
+  // sure we don't create a new state for the JavaThread.
```

Comment:
I think it is important to maintain `JvmtiThreadState::_thread_oop_h` correctly for the attached native thread. With the proposed change, `JvmtiThreadState::_thread_oop_h` could stay null for an attached native thread, which seems wrong.

It may be OK since `JvmtiThreadState::_thread_oop_h` is only used by support for virtual threads. It is unlikely that an attached native thread becomes a carrier for a virtual thread. However, it is probably still desirable to update `JvmtiThreadState::_thread_oop_h` to the correct java.lang.Thread oop.

Reply:
Thanks for the input @caoman. I updated the PR to avoid creating a JvmtiThreadState during attaching and allocating thread oop. I think that avoids a incomplete JvmtiThreadState being created, which is seems to be cleaner.